### DISCUSSION TITLE
NeuralNet Tracker Pitch Rotation Correction

### DISCUSSION THREAD

**Discussion by minimufffin** [opentrack/opentrack#1522] [Q&A]
@DaWelter I assume the below code accounts for yaw changing when your head is just translating. 

mat33 compute_rotation_correction(const cv::Point2f &p, float focal_length_x)
{
    return rotation_from_two_vectors(
          {1.f,0.f,0.f}, 
          {focal_length_x, p.y, p.x});
}

Is there a way to modify this to also account for change in pitch with translation, using focal_length_y? I don't fully understand the rotation matrix but I think it should be possible using the same concept.

**Comment by DaWelter** [comment]
It should do the pitch already. The idea is to compute the 3d rotation from the forward vector (1,0,0) to the direction vector pointing toward the face center. So if you move up/down it should result in a pitch correction.

I don't  recall why I put focal_length_x there and not a 1. I hope this code does not break with different aspect ratios. My comments are also not quite right because the first component of the vector points into the screen.

I will keep in mind to revisit this part.

Hope this helps

**Reply by minimufffin** [reply]
Thanks for the clarification, I think I understand better now.

Wouldn't you still need to know the FOV of the camera to compute the rotation correction? Then you would know what angle to offset by at the edges of the screen. This is what I came up with to compute the correction:
```
mat33 compute_rotation_correction(const cv::Point2f &p, float focal_length_x, float focal_length_y)
{
    return rotation_from_two_vectors(
        {1.f,0.f,0.f}, 
        {1.f, std::tan(p.y * std::atan(1 / focal_length_y)), std::tan(p.x * std::atan(1 / focal_length_x))});
}
```
atan(1/focal_length) is basically the FOV along an axis. 

If the focal lengths are greater than 1 then the following is a good approximation:
```
mat33 compute_rotation_correction(const cv::Point2f &p, float focal_length_x, float focal_length_y)
{
    return rotation_from_two_vectors(
        {1.f,0.f,0.f}, 
        {1.f, p.y / focal_length_y, p.x / focal_length_x});
}
```
It is hard to say if this change is any better, this and the original generally seem to be working for me.
